
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.46">
    
    
      
        <title>Introduction to Tufts HPC Cluster - Tufts TTS Research Technology Bioinformatics Workshops</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      
  
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
    
    
  
  
  <style>:root{--md-admonition-icon--note:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1 7.775V2.75C1 1.784 1.784 1 2.75 1h5.025c.464 0 .91.184 1.238.513l6.25 6.25a1.75 1.75 0 0 1 0 2.474l-5.026 5.026a1.75 1.75 0 0 1-2.474 0l-6.25-6.25A1.75 1.75 0 0 1 1 7.775m1.5 0c0 .066.026.13.073.177l6.25 6.25a.25.25 0 0 0 .354 0l5.025-5.025a.25.25 0 0 0 0-.354l-6.25-6.25a.25.25 0 0 0-.177-.073H2.75a.25.25 0 0 0-.25.25ZM6 5a1 1 0 1 1 0 2 1 1 0 0 1 0-2"/></svg>');--md-admonition-icon--abstract:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>');--md-admonition-icon--info:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8m8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13M6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75M8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2"/></svg>');--md-admonition-icon--tip:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M3.499.75a.75.75 0 0 1 1.5 0v.996C5.9 2.903 6.793 3.65 7.662 4.376l.24.202c-.036-.694.055-1.422.426-2.163C9.1.873 10.794-.045 12.622.26 14.408.558 16 1.94 16 4.25c0 1.278-.954 2.575-2.44 2.734l.146.508.065.22c.203.701.412 1.455.476 2.226.142 1.707-.4 3.03-1.487 3.898C11.714 14.671 10.27 15 8.75 15h-6a.75.75 0 0 1 0-1.5h1.376a4.5 4.5 0 0 1-.563-1.191 3.84 3.84 0 0 1-.05-2.063 4.65 4.65 0 0 1-2.025-.293.75.75 0 0 1 .525-1.406c1.357.507 2.376-.006 2.698-.318l.009-.01a.747.747 0 0 1 1.06 0 .75.75 0 0 1-.012 1.074c-.912.92-.992 1.835-.768 2.586.221.74.745 1.337 1.196 1.621H8.75c1.343 0 2.398-.296 3.074-.836.635-.507 1.036-1.31.928-2.602-.05-.603-.216-1.224-.422-1.93l-.064-.221c-.12-.407-.246-.84-.353-1.29a2.4 2.4 0 0 1-.507-.441 3.1 3.1 0 0 1-.633-1.248.75.75 0 0 1 1.455-.364c.046.185.144.436.31.627.146.168.353.305.712.305.738 0 1.25-.615 1.25-1.25 0-1.47-.95-2.315-2.123-2.51-1.172-.196-2.227.387-2.706 1.345-.46.92-.27 1.774.019 3.062l.042.19.01.05c.348.443.666.949.94 1.553a.75.75 0 1 1-1.365.62c-.553-1.217-1.32-1.94-2.3-2.768L6.7 5.527c-.814-.68-1.75-1.462-2.692-2.619a3.7 3.7 0 0 0-1.023.88c-.406.495-.663 1.036-.722 1.508.116.122.306.21.591.239.388.038.797-.06 1.032-.19a.75.75 0 0 1 .728 1.31c-.515.287-1.23.439-1.906.373-.682-.067-1.473-.38-1.879-1.193L.75 5.677V5.5c0-.984.48-1.94 1.077-2.664.46-.559 1.05-1.055 1.673-1.353z"/></svg>');--md-admonition-icon--success:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.75.75 0 0 1 .018-1.042.75.75 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0"/></svg>');--md-admonition-icon--question:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8m8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13M6.92 6.085h.001a.749.749 0 1 1-1.342-.67c.169-.339.436-.701.849-.977C6.845 4.16 7.369 4 8 4a2.76 2.76 0 0 1 1.637.525c.503.377.863.965.863 1.725 0 .448-.115.83-.329 1.15-.205.307-.47.513-.692.662-.109.072-.22.138-.313.195l-.006.004a6 6 0 0 0-.26.16 1 1 0 0 0-.276.245.75.75 0 0 1-1.248-.832c.184-.264.42-.489.692-.661q.154-.1.313-.195l.007-.004c.1-.061.182-.11.258-.161a1 1 0 0 0 .277-.245C8.96 6.514 9 6.427 9 6.25a.61.61 0 0 0-.262-.525A1.27 1.27 0 0 0 8 5.5c-.369 0-.595.09-.74.187a1 1 0 0 0-.34.398M9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0"/></svg>');--md-admonition-icon--warning:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0M9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0"/></svg>');--md-admonition-icon--failure:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.344 2.343za8 8 0 0 1 11.314 11.314A8.002 8.002 0 0 1 .234 10.089a8 8 0 0 1 2.11-7.746m1.06 10.253a6.5 6.5 0 1 0 9.108-9.275 6.5 6.5 0 0 0-9.108 9.275M6.03 4.97 8 6.94l1.97-1.97a.749.749 0 0 1 1.275.326.75.75 0 0 1-.215.734L9.06 8l1.97 1.97a.749.749 0 0 1-.326 1.275.75.75 0 0 1-.734-.215L8 9.06l-1.97 1.97a.749.749 0 0 1-1.275-.326.75.75 0 0 1 .215-.734L6.94 8 4.97 6.03a.75.75 0 0 1 .018-1.042.75.75 0 0 1 1.042-.018"/></svg>');--md-admonition-icon--danger:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M9.504.43a1.516 1.516 0 0 1 2.437 1.713L10.415 5.5h2.123c1.57 0 2.346 1.909 1.22 3.004l-7.34 7.142a1.25 1.25 0 0 1-.871.354h-.302a1.25 1.25 0 0 1-1.157-1.723L5.633 10.5H3.462c-1.57 0-2.346-1.909-1.22-3.004zm1.047 1.074L3.286 8.571A.25.25 0 0 0 3.462 9H6.75a.75.75 0 0 1 .694 1.034l-1.713 4.188 6.982-6.793A.25.25 0 0 0 12.538 7H9.25a.75.75 0 0 1-.683-1.06l2.008-4.418.003-.006-.004-.009-.006-.006-.008-.001q-.005 0-.009.004"/></svg>');--md-admonition-icon--bug:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M4.72.22a.75.75 0 0 1 1.06 0l1 .999a3.5 3.5 0 0 1 2.441 0l.999-1a.748.748 0 0 1 1.265.332.75.75 0 0 1-.205.729l-.775.776c.616.63.995 1.493.995 2.444v.327q0 .15-.025.292c.408.14.764.392 1.029.722l1.968-.787a.75.75 0 0 1 .556 1.392L13 7.258V9h2.25a.75.75 0 0 1 0 1.5H13v.5q-.002.615-.141 1.186l2.17.868a.75.75 0 0 1-.557 1.392l-2.184-.873A5 5 0 0 1 8 16a5 5 0 0 1-4.288-2.427l-2.183.873a.75.75 0 0 1-.558-1.392l2.17-.868A5 5 0 0 1 3 11v-.5H.75a.75.75 0 0 1 0-1.5H3V7.258L.971 6.446a.75.75 0 0 1 .558-1.392l1.967.787c.265-.33.62-.583 1.03-.722a1.7 1.7 0 0 1-.026-.292V4.5c0-.951.38-1.814.995-2.444L4.72 1.28a.75.75 0 0 1 0-1.06m.53 6.28a.75.75 0 0 0-.75.75V11a3.5 3.5 0 1 0 7 0V7.25a.75.75 0 0 0-.75-.75ZM6.173 5h3.654A.17.17 0 0 0 10 4.827V4.5a2 2 0 1 0-4 0v.327c0 .096.077.173.173.173"/></svg>');--md-admonition-icon--example:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.75.75 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--prerequisite:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 5.782V2.5h-.25a.75.75 0 0 1 0-1.5h6.5a.75.75 0 0 1 0 1.5H11v3.282l3.666 5.76C15.619 13.04 14.543 15 12.767 15H3.233c-1.776 0-2.852-1.96-1.899-3.458Zm-2.4 6.565a.75.75 0 0 0 .633 1.153h9.534a.75.75 0 0 0 .633-1.153L12.225 10.5h-8.45ZM9.5 2.5h-3V6c0 .143-.04.283-.117.403L4.73 9h6.54L9.617 6.403A.75.75 0 0 1 9.5 6Z"/></svg>');--md-admonition-icon--quote:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M1.75 2.5h10.5a.75.75 0 0 1 0 1.5H1.75a.75.75 0 0 1 0-1.5m4 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5m0 5h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1 0-1.5M2.5 7.75v6a.75.75 0 0 1-1.5 0v-6a.75.75 0 0 1 1.5 0"/></svg>');}</style>



    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#introduction-to-tufts-hpc-cluster" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Tufts TTS Research Technology Bioinformatics Workshops" class="md-header__button md-logo" aria-label="Tufts TTS Research Technology Bioinformatics Workshops" data-md-component="logo">
      
  <img src="../../img/logo2.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Tufts TTS Research Technology Bioinformatics Workshops
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Introduction to Tufts HPC Cluster
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue" data-md-color-accent="blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/tuftsdatalab/tuftsWorkshops" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  About

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../2022_workshops/2022_workshops/" class="md-tabs__link">
          
  
  2022 Workshops

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../2023_workshops/2023_workshops/" class="md-tabs__link">
          
  
  2023 Workshops

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../readme/" class="md-tabs__link">
          
  
  2024 Workshops

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Tufts TTS Research Technology Bioinformatics Workshops" class="md-nav__button md-logo" aria-label="Tufts TTS Research Technology Bioinformatics Workshops" data-md-component="logo">
      
  <img src="../../img/logo2.png" alt="logo">

    </a>
    Tufts TTS Research Technology Bioinformatics Workshops
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tuftsdatalab/tuftsWorkshops" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    About
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    2022 Workshops
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            2022 Workshops
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/2022_workshops/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Fall 2022
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            Fall 2022
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_1" >
        
          
          <label class="md-nav__link" for="__nav_2_2_1" id="__nav_2_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Intro to HPC/Slurm For Life Sciences
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_1">
            <span class="md-nav__icon md-icon"></span>
            Intro to HPC/Slurm For Life Sciences
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/01_hpc-intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/02_request-account/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cluster Account and Storage
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/03_login/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Login
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/04_file-transfer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    File Transfer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/05_what-is-the-cluster/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Terninologies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/06_compute-resources/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Compute Resources
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/07_software-cluster/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Software on the Cluster
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/08_tmux/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tmux Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/09_interactive-session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Interactive Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/10_job-status/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Job Status
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/11_intro-to-unix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intro To Linux
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/12_start-w-shell/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Starting with Shell
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/13_bash-parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Bash Parameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/14_shell-navigation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Shell Navigation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/15_create-manipulate-files/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating & Manipulating Files
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/16_going-home/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Going Home
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/17_running-interactive/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running an Interactive Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/18_blast-example/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Example with BLAST
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introHpcSlurm/19_blast-batch/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    BLAST Batch Script
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2_2" id="__nav_2_2_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Intro to RStudio For Life Sciences
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_2">
            <span class="md-nav__icon md-icon"></span>
            Intro to RStudio For Life Sciences
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introRStudio/01_r-intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introRStudio/02_r-ondemand/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introRStudio/03_r-workshop-markdown/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intro To R Script
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_3" >
        
          
          <label class="md-nav__link" for="__nav_2_2_3" id="__nav_2_2_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Intro To 16S Metabarcoding
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_3">
            <span class="md-nav__icon md-icon"></span>
            Intro To 16S Metabarcoding
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/01_metabarcoding-intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/02_background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Background
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/03_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/04_quality-control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/05_error-asv/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Error Model & ASVs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/06_merging-chimeras-taxonomy/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Merging, Chimeras & Taxonomy
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/07_diversity-analysis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Diversity Analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/intro16S/08_differential-abundance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Differential Abundance
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_4" >
        
          
          <label class="md-nav__link" for="__nav_2_2_4" id="__nav_2_2_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Intro To Metagenomics
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_4">
            <span class="md-nav__icon md-icon"></span>
            Intro To Metagenomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introMetagenomics/01_intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introMetagenomics/02_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introMetagenomics/03_Metagenomics.nb.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Workshop Notebook
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_5" >
        
          
          <label class="md-nav__link" for="__nav_2_2_5" id="__nav_2_2_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Intro To Genome Technology
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_5">
            <span class="md-nav__icon md-icon"></span>
            Intro To Genome Technology
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introGenomeTech/01_intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introGenomeTech/Rhodes_BioE291.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Slides
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_6" >
        
          
          <label class="md-nav__link" for="__nav_2_2_6" id="__nav_2_2_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Intro To AlphaFold2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_6">
            <span class="md-nav__icon md-icon"></span>
            Intro To AlphaFold2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/01_background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Background
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/02_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/03_af2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AlphaFold2 Algorithm
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/05_af2_output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AlphaFold2 Output
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/06_pymol_viz/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyMOL Visualization
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2022_workshops/introAlphaFold/07_af2_batch/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Optional: AlphaFold2 Batch Script
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    2023 Workshops
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            2023 Workshops
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/2023_workshops/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Spring 2023
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Spring 2023
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_1" >
        
          
          <label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Finding & Analyzing Metagenomics Data
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_1">
            <span class="md-nav__icon md-icon"></span>
            Finding & Analyzing Metagenomics Data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/metagenomeData/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/metagenomeData/01_searchSRA/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Searching For SRA Data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/metagenomeData/02_stat/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    STAT Taxonomic Analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/metagenomeData/03_kraken/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Kraken2 Taxonomic Analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/metagenomeData/04_compareStatKraken/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Compare STAT & Kraken2
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    RNA-Seq with Galaxy
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_2">
            <span class="md-nav__icon md-icon"></span>
            RNA-Seq with Galaxy
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/galaxyRNASeq/00_Galaxy_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/galaxyRNASeq/01_Introduction_and_Setup_noqual/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Processing Raw Reads
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/galaxyRNASeq/03_Read_alignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read Alignment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/galaxyRNASeq/04_Gene_quantification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Gene Quantification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/galaxyRNASeq/05_Diff_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Differential Expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Cas12a Variant Structural Predictions With AlphaFold2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3">
            <span class="md-nav__icon md-icon"></span>
            Cas12a Variant Structural Predictions With AlphaFold2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/cas12aAlphaFold2/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/cas12aAlphaFold2/01_CRISPR-Cas_Structural_Predictions_with_AlphaFold2.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction To Cas12a Variant Structural Predictions With AlphaFold2
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/cas12aAlphaFold2/02_AlphaFold2_Accuracy_Assessment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    AlphaFold2 Accuracy Assessment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/cas12aAlphaFold2/03_Cas12a_Variant_Vizualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cas12a Variant Visualization
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_4" >
        
          
          <label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Trajectory Analysis
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_4">
            <span class="md-nav__icon md-icon"></span>
            Trajectory Analysis
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/trajectoryAnalysis/00_background/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/trajectoryAnalysis/01_setup_and_monocle/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup and Monocle3 Objects
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/trajectoryAnalysis/02_monocle_setup_and_trajectory_analysis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Monocle3 Workflow and Trajectory Analysis
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../2023_workshops/trajectoryAnalysis/03_pseudotime_and_differential_expression/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pseudotime and Differential Expression
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    2024 Workshops
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            2024 Workshops
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Spring 2024
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Spring 2024
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2_1" >
        
          
          <label class="md-nav__link" for="__nav_4_2_1" id="__nav_4_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Introduction to Metagenomics
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2_1">
            <span class="md-nav__icon md-icon"></span>
            Introduction to Metagenomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../IGDH-1001_2024Feb/Hands-on_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    In class hands-on session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../IGDH-1001_2024Feb/Take-home_exercise/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Take-home exercise
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../IGDH-1001_2024Feb/Take-home_exercise_announcement/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Take-home exercise announcement
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../IGDH-1001_2024Feb/Exercise2_hints.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exercise Hint
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../IGDH-1001_2024Feb/Exercise3_Example_Report.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Example Report
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2_2" id="__nav_4_2_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Introduction to AlphaFold2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2_2">
            <span class="md-nav__icon md-icon"></span>
            Introduction to AlphaFold2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cas12aAlphaFold2_sp24/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cas12aAlphaFold2_sp24/01_Run_AlphaFold2_CLI/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run AlphaFold2_CLI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cas12aAlphaFold2_sp24/02_Run_AlphaFold2_OpenOndemandApp/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run AlphaFold2 OpenOndemandApp
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cas12aAlphaFold2_sp24/03_Vizualize_predicted_structure_with_PYMOL/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vizualize with PYMOL
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cas12aAlphaFold2_sp24/Alphafold_ShirleyLi_CHBE0165_2024-03.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Slides
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2_3" >
        
          
          <label class="md-nav__link" for="__nav_4_2_3" id="__nav_4_2_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Running RNA-Seq analysis with nextflow & nf-core
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2_3">
            <span class="md-nav__icon md-icon"></span>
            Running RNA-Seq analysis with nextflow & nf-core
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../nfcore_rnaseq_sp24/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../nfcore_rnaseq_sp24/01_fetchngs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    nf-core fetchngs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../nfcore_rnaseq_sp24/02_rnaseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    nf-core rnaseq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../nfcore_rnaseq_sp24/03_differentialabundance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    nf-core differentialabundance
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../nfcore_rnaseq_sp24/04_multiqc_report.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    multiqc report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../nfcore_rnaseq_sp24/05_differentialabundance_report.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    differentialabundance report
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Fall 2024
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Fall 2024
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics_workshop_agenda/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Workshop Agenda
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3_2" >
        
          
          <label class="md-nav__link" for="__nav_4_3_2" id="__nav_4_3_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    1. Introduction to Tufts HPC Cluster
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3_2">
            <span class="md-nav__icon md-icon"></span>
            1. Introduction to Tufts HPC Cluster
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="Intro2HPC/readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Tufts HPC Cluster
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="Intro2HPC/00_Tufts_HPC_Terminologies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tufts HPC Terminologies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="Intro2HPC/01_Getting_to_Know_Tufts_HPC_Cluster_Resources/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Getting to Know Tufts HPC Cluster Resources
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="Intro2HPC/02_Tufts_HPC_Cluster_Access/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tufts HPC Cluster Access
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="Intro2HPC/03_Getting_Work_Done_on_Tufts_HPC_Cluster_with_SLURM/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Getting Work Done on Tufts HPC Cluster with SLURM
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3_3" id="__nav_4_3_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    2. Getting started with Bioinformatics on Tufts HPC
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3_3">
            <span class="md-nav__icon md-icon"></span>
            2. Getting started with Bioinformatics on Tufts HPC
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3_3_1" >
        
          
          <label class="md-nav__link" for="__nav_4_3_3_1" id="__nav_4_3_3_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Linux/Unix command line basics
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_3_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3_3_1">
            <span class="md-nav__icon md-icon"></span>
            Linux/Unix command line basics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/linux/00_overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/linux/01_what_is_linux/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Linux/Unix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/linux/02_files/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Files and File system
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/linux/03_basictools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Must-known Linux/Unix tools
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/linux/04_loops/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Conditional statements and loops
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/linux/05_advanced_tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Advanced Linux/Unix tools
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/02_tufts_bioinformatics_resources/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Available bioinformatics resources/tools on Tufts HPC
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics201/03_run_bioinformatics_on_HPC/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to run bioinformatics analysis on Tufts HPC
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3_4" >
        
          
          <label class="md-nav__link" for="__nav_4_3_4" id="__nav_4_3_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    3. Software installation and environment management in HPC
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3_4">
            <span class="md-nav__icon md-icon"></span>
            3. Software installation and environment management in HPC
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics301/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics301/01_source/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Application Installation from Source Codes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics301/02_Rpackage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    R package Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics301/03_conda/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Conda Environment Management, Jupyer Kernel and Modules
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics301/04_conda-env-mod/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Simplying Conda Environment Management with conda-env-mod
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3_5" >
        
          
          <label class="md-nav__link" for="__nav_4_3_5" id="__nav_4_3_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    4. Advanced HPC workflows and job management
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3_5">
            <span class="md-nav__icon md-icon"></span>
            4. Advanced HPC workflows and job management
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics401/00_introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics401/01_array_jobs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Array jobs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics401/02_nextflow_tufts-2024Fall.pdf" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Nextflow Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../2024_bioinformatics401/03_nfcore/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    How to run nf-core pipelines
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tts-research-technology" class="md-nav__link">
    <span class="md-ellipsis">
      TTS - Research Technology
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/tuftsdatalab/tuftsWorkshops/blob/main/docs/2024_workshops/2024_bioinformatics101.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<h1 id="introduction-to-tufts-hpc-cluster">Introduction to Tufts HPC Cluster</h1>
<blockquote>
<p>Date: 10/09/2024</p>
<p>Delilah Maloney
Sr. HPC Specialist</p>
<h6 id="tts-research-technology">TTS - Research Technology</h6>
</blockquote>
<p>Questions, Issues, Requests? </p>
<p><strong>Contact us at: tts-research@tufts.edu</strong></p>
<p><strong><a href="https://it.tufts.edu/researchtechnology.tufts.edu">Research Technology Website</a></strong></p>
<div class="toc">
<ul>
<li><a href="#introduction-to-tufts-hpc-cluster">Introduction to Tufts HPC Cluster</a><ul>
<li><a href="#tts-research-technology">TTS - Research Technology</a></li>
</ul>
</li>
<li><a href="#tufts-hpc-terminologies">Tufts HPC Terminologies</a><ul>
<li><a href="#what-is-hpc">What is "HPC"</a></li>
<li><a href="#what-is-cluster">What is "Cluster"</a></li>
<li><a href="#cpu-vs-gpu">CPU vs GPU</a></li>
<li><a href="#cores-vs-node">Cores vs Node</a></li>
<li><a href="#memory-vs-storage">Memory vs Storage</a></li>
</ul>
</li>
<li><a href="#getting-to-know-tufts-hpc-cluster-resources">Getting to Know Tufts HPC Cluster Resources</a><ul>
<li><a href="#account-storage-requests">Account &amp; Storage Requests</a></li>
<li><a href="#tufts-hpc-cluster">Tufts HPC Cluster</a></li>
<li><a href="#cluster-storage">Cluster Storage</a></li>
<li><a href="#cpus">CPUs</a></li>
<li><a href="#gpus">GPUs</a></li>
<li><a href="#cluster-resource-limit">Cluster Resource Limit</a></li>
<li><a href="#cluster-computing-resource-availability">Cluster Computing Resource Availability</a></li>
</ul>
</li>
<li><a href="#tufts-hpc-cluster-access">Tufts HPC Cluster Access</a><ul>
<li><a href="#ondemand">OnDemand</a></li>
<li><a href="#shell-access">Shell Access</a><ul>
<li><a href="#mac-osx-linux">Mac OSX &amp; Linux</a></li>
<li><a href="#windows">Windows</a></li>
</ul>
</li>
<li><a href="#file-transfers-tofrom-tufts-hpc-cluster">File Transfers To/From Tufts HPC Cluster</a><ul>
<li><a href="#file-transfer-clients">File Transfer Clients</a></li>
<li><a href="#ondemand_1">OnDemand</a></li>
<li><a href="#terminal">Terminal</a></li>
<li><a href="#globus-coming-soon">Globus - Coming Soon!</a></li>
</ul>
</li>
<li><a href="#software-modules-packages">Software, Modules, Packages</a><ul>
<li><a href="#what-are-modules">What are modules?</a></li>
<li><a href="#whats-new">What's NEW?</a></li>
<li><a href="#how-to-use-modules">How to use modules?</a></li>
<li><a href="#install-softwarepackages">Install Software/Packages</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#getting-work-done-on-tufts-hpc-cluster-with-slurm">Getting Work Done on Tufts HPC Cluster with SLURM</a><ul>
<li><a href="#ondemand_2">OnDemand</a></li>
<li><a href="#slurm-information">Slurm Information</a></li>
<li><a href="#utilize-new-hpctools-module">Utilize New hpctools Module !!!</a></li>
<li><a href="#batch-job">Batch Job</a></li>
<li><a href="#job-status">Job Status</a></li>
</ul>
</li>
</ul>
</div>
<hr />
<h1 id="tufts-hpc-terminologies">Tufts HPC Terminologies</h1>
<h2 id="what-is-hpc">What is "HPC"</h2>
<ul>
<li>
<p>High-performance computing (HPC) is the ability to process data and perform complex calculations at high speeds.</p>
</li>
<li>
<p>Everything is relative</p>
</li>
</ul>
<h2 id="what-is-cluster">What is "Cluster"</h2>
<ul>
<li>A computer cluster is a set of loosely or tightly <strong>connected</strong> <strong>computers</strong> that work together so that, in many respects, they can be viewed as a single system. Computer clusters have each node set to perform the same/similar tasks, controlled and scheduled by <strong><a href="#SLURM">software</a></strong>. </li>
</ul>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/Cluster.png" alt="Cluster" width=70%></p>
<p><strong><em>To parallel, or not to parallel? That is the question.</em></strong></p>
<h2 id="cpu-vs-gpu">CPU vs GPU</h2>
<ul>
<li>
<p>CPU -- <strong>Central Processing Unit</strong> </p>
</li>
<li>
<p>A CPU can never be fully replaced by a GPU</p>
</li>
<li>
<p>Can be thought of as the <strong>taskmaster</strong> of the entire system, coordinating a wide range of general-purpose computing tasks</p>
</li>
<li>
<p>GPU -- <strong>Graphics Processing Unit</strong></p>
</li>
<li>
<p>GPUs were originally designed to create images for computer graphics and video game consoles</p>
</li>
<li>
<p>GPGPU</p>
</li>
<li>
<p>Performing a narrower range of more specialized tasks</p>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/CPUGPU.png" alt="CPU-GPU" width=60%></p>
</li>
</ul>
<h2 id="cores-vs-node">Cores vs Node</h2>
<ul>
<li>A <strong>node</strong> is a single computer in the system, which has a number of computing units, <strong>cores</strong>. </li>
</ul>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/CoreNode.png" alt="Core-Node" width=70%></p>
<h2 id="memory-vs-storage">Memory vs Storage</h2>
<p>The central processing unit (CPU) of a computer is what manipulates data by performing computations.  In practice, almost all computers use a <a href="https://en.wikipedia.org/wiki/Memory_hierarchy">storage hierarchy</a>, which puts fast but expensive and small storage options close to the  CPU and slower but less expensive and larger options further away.  Generally the fast volatile technologies (which lose data when off  power) are referred to as "memory", while slower persistent technologies are referred to as "storage".</p>
<ul>
<li>Memory</li>
<li>Small, fast, expensive</li>
<li>Used to store information for immediate use</li>
<li>Volatile.</li>
<li>Storage</li>
<li>Larger, slower, cheaper</li>
<li>Non-volatile (retaining data when its power is shut off)</li>
</ul>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/MemStorage.png" alt="Memory-Storage" width=60%></p>
<h1 id="getting-to-know-tufts-hpc-cluster-resources">Getting to Know Tufts HPC Cluster Resources</h1>
<h2 id="account-storage-requests">Account &amp; Storage Requests</h2>
<p>Go to <strong><a href="https://it.tufts.edu/high-performance-computing">Tufts HPC website</a></strong> for HPC Cluster Account and Storage Requests</p>
<p>For Faculties ONLY: <strong><a href="https://tufts.qualtrics.com/jfe/form/SV_d7o0UZFgK1PFXnv">HPC for Classes</a></strong></p>
<h2 id="tufts-hpc-cluster">Tufts HPC Cluster</h2>
<ul>
<li>Moved to <strong><a href="https://www.mghpcc.org/">MGHPCC</a></strong> in January 2024!</li>
</ul>
<h2 id="cluster-storage">Cluster Storage</h2>
<ul>
<li><strong>Home Directory</strong></li>
</ul>
<p>Every user has a home directory.</p>
<p>Be aware! Your Home Directory (<strong>30GB</strong>, fixed) should be <code>/cluster/home/your_utln</code></p>
<p>If you are not sure how much storage you have used in your home directory, feel free to contact us and we can provide you the number. </p>
<p>For self-service, you can use the following commands from a shell terminal to find out your home directory usage:</p>
<p><code>$ module load hpctools</code></p>
<p><code>$ hpctools</code> (from any node) </p>
<p>OR </p>
<p><code>$ du -a -h --max-depth=1 ~ | sort -hr</code> from a <strong>compute node</strong> in your home directory to find out the detailed usage. </p>
<ul>
<li><strong>Reserach Project Storage</strong></li>
</ul>
<p><strong><a href="https://it.tufts.edu/research-technology/">Request research project storage</a></strong></p>
<p>Created for research labs and classes. A Tufts HPC cluster research project storage share can <strong>only</strong> be owned by a <strong>Tufts</strong> <strong>faculty</strong>.</p>
<p>New Storage Policies can be found on <a href="https://it.tufts.edu/research-technology/announcements">RT Announcements</a> page. - Tiered Storage</p>
<p>Your research projet storage (from <strong>50GB</strong>) path should be <code>/cluster/tufts/yourlabname/</code>, and each member of the lab group has a dedicated directory <code>/cluster/tufts/yourlabname/your_utln</code></p>
<p>To see your <strong>research project storage quota</strong> by running the following command from <strong>any node on the new cluster Pax</strong>:</p>
<p><code>$ df -h /cluster/tufts/yourlabname</code> </p>
<p>OR </p>
<p><code>$ module load hpctools</code></p>
<p><code>$ hpctools</code></p>
<p><strong>NOTE:</strong> Accessing your research project storage space for the <strong>first time</strong> in your current session, please make sure you type out the <strong>FULL PATH</strong> to the directory <code>/cluster/tufts/yourlabname/</code>.</p>
<h2 id="cpus">CPUs</h2>
<p>Primarily Intel Xeon CPUs, from Broadwell to Emerald Rapids, with hyperthreading enabled*.</p>
<p>Compute nodes are grouped into <strong>partitions</strong> based on <u>functionality</u> and <u>priority</u> levels.</p>
<p><strong>Public Partitions :</strong></p>
<div class="highlight"><pre><span></span><code>PARTITION       TIMELIMIT      
batch*          7-00:00:00          
gpu             7-00:00:00        
interactive     4:00:00        
largemem        7-00:00:00        
mpi             7-00:00:00         
preempt         7-00:00:00     
</code></pre></div>
<ul>
<li><strong>preempt</strong> - Be aware, the <code>preempt</code> partition consists of most of the nodes on the cluster, including public nodes and <strong>contrib nodes</strong> from different research labs. When submitting jobs to preempt partition, you acknowledge that your jobs are taking the <strong>risk</strong> of being preempted by higher priority jobs. In that case, you will simply have to resubmit your jobs. </li>
</ul>
<h2 id="gpus">GPUs</h2>
<p><strong>NVIDIA GPUs</strong> are available in <code>gpu</code> and <code>preempt</code> partitions</p>
<ul>
<li>
<p><strong>Request GPU resources with <code>--gres</code>. See details below.</strong></p>
</li>
<li>
<p>If no specific architecture is required, GPU resources can be request with<code>--gres=gpu:1</code> (one GPU)</p>
</li>
<li>
<p>You can request more than one type of GPUs with <code>constraint</code>, e.g.  </p>
</li>
</ul>
<p><code>--gres=gpu:1 --constraint="t4|p100|v100"</code></p>
<ul>
<li>
<p>Please <strong>DO NOT</strong> manually set <code>CUDA_VISIBLE_DEVICES</code>. </p>
</li>
<li>
<p>Users can ONLY see GPU devices that are assigned to them with <code>$ nvidia-smi</code>.</p>
</li>
</ul>
<p>If you submit batch jobs, it's recommended adding <code>nvidia-smi</code> in your slurm job submission script.</p>
<ul>
<li>
<p><code>gpu</code> partition<code>-p gpu</code>:</p>
</li>
<li>
<p>NVIDIA A100</p>
<ul>
<li>In "gpu" partition</li>
<li>Request with: <code>--gres=gpu:a100:1</code>(one A100 GPU, can request up to 8 on one node)</li>
<li><code>--constraint="a100-80G"</code></li>
<li>Each GPU comes with 80GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA P100s</p>
<ul>
<li>In "gpu" partition</li>
<li>Request with: <code>--gres=gpu:p100:1</code>(one P100 GPU, can request up to 4 on one node)</li>
<li><code>--constraint="p100"</code></li>
<li>Each GPU comes with 16GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p><code>preempt</code> partition <code>-p preempt</code>:</p>
</li>
<li>
<p><code>a100</code>, <code>v100</code>, <code>p100</code>, <code>rtx_6000</code>, <code>rtx_a6000</code>, <code>rtx_6000ada</code>, <code>rtx_a5000</code>, <code>h100</code>, <code>l40s</code>, <code>t4</code></p>
</li>
<li>
<p>NVIDIA T4</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:t4:1</code>(one T4 GPU, can request up to 4 on one node)</li>
<li><code>--constraint="t4"</code></li>
<li>Each GPU comes with 16GB of DRAM</li>
<li>Driver supports upto CUDA 10.2</li>
</ul>
</li>
<li>
<p>NVIDIA P100</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:p100:1</code>(one P100 GPU, can request up to 6 on one node)</li>
<li><code>--constraint="p100"</code></li>
<li>Each GPU comes with 16GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA rtx_6000</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:rtx_6000:1</code>(one RTX_6000 GPU, can request up to 8 on one node)</li>
<li><code>--constraint="rtx_6000"</code></li>
<li>Each GPU comes with 24GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA rtx_a6000</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:rtx_a6000:1</code>(one RTX_A6000 GPU, can request up to 8 on one node)</li>
<li><code>--constraint="rtx_a6000"</code></li>
<li>Each GPU comes with 48GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA rtx_6000ada</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:rtx_6000ada:1</code>(one RTX_6000Ada GPU, can request up to 4 on one node)</li>
<li><code>--constraint="rtx_6000ada"</code></li>
<li>Each GPU comes with 48GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA V100</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:v100:1</code>(one V100 GPU, can request up to 4 on one node)</li>
<li><code>--constraint="v100"</code></li>
<li>Each GPU comes with 16GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA A100</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:a100:1</code>(one A100 GPU, can request up to 8 on one node)</li>
<li><code>--constraint="a100-80G"</code></li>
<li><code>--constraint="a100-40G"</code></li>
<li><code>--constraint="a100"</code></li>
<li>Each GPU comes with 40GB of DRAM or 80GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA H100</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:h100:1</code>(one V100 GPU, can request up to 3 on one node)</li>
<li><code>--constraint="h100"</code></li>
<li>Each GPU comes with 80GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA L40s</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:l40s:1</code>(one L40s GPU, can request up to 4 on one node)</li>
<li><code>--constraint="l40s"</code></li>
<li>Each GPU comes with 48GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA RTX A5000</p>
<ul>
<li>In "preempt" partition</li>
<li>Request with: <code>--gres=gpu:rtx_a5000:1</code>(one RTX A5000 GPU, can request up to 4 on one node)</li>
<li><code>--constraint="l40s"</code></li>
<li>Each GPU comes with 48GB of DRAM</li>
<li>Driver supports upto CUDA 12.2</li>
</ul>
</li>
<li>
<p>NVIDIA L40</p>
<ul>
<li>
<p>In "preempt" partition</p>
</li>
<li>
<p>Request with: <code>--gres=gpu:l40:1</code>(one L40 GPU, can request up to 4 on one node)</p>
</li>
<li>
<p><code>--constraint="l40"</code></p>
</li>
<li>
<p>Each GPU comes with 48GB of DRAM</p>
</li>
<li>
<p>Driver supports upto CUDA 12.2</p>
</li>
</ul>
</li>
</ul>
<h2 id="cluster-resource-limit">Cluster Resource Limit</h2>
<ul>
<li>
<p><strong>Public Partitions</strong> (batch+mpi+largemem+gpu) </p>
</li>
<li>
<p>CPU: 500 cores</p>
<p>RAM: 2000 GB</p>
<p>GPU: 5</p>
</li>
<li>
<p><strong>Preempt Partition</strong> (preempt) </p>
</li>
<li>
<p>CPU: 1000 cores</p>
<p>RAM: 4000 GB</p>
<p>GPU: 10</p>
</li>
<li>
<p><strong>Maximun Number of Jobs submitted</strong>: 1000</p>
</li>
</ul>
<h2 id="cluster-computing-resource-availability">Cluster Computing Resource Availability</h2>
<p><code>$ module load hpctools</code></p>
<p><code>$ hpctools</code> </p>
<p>Then follow the on-screen instructions to extract the information you need. </p>
<hr />
<h1 id="tufts-hpc-cluster-access"><strong>Tufts HPC Cluster Access</strong></h1>
<hr />
<blockquote>
<p><strong>VPN</strong></p>
<ul>
<li>Off-campus Non-Tufts Network please connect to <a href="https://access.tufts.edu/vpn">Tufts VPN</a></li>
</ul>
<p><strong>2FA</strong></p>
<ul>
<li>DUO is needed on Tufts Network (not needed for <a href="https://ondemand.pax.tufts.edu"><strong>OnDemand</strong></a>, https://ondemand.pax.tufts.edu)</li>
<li>DUO is needed when using FastX11 from  <a href="https://ondemand.pax.tufts.edu"><strong>OnDemand</strong></a></li>
</ul>
<p><strong>SSH</strong></p>
<ul>
<li>The SSH protocol aka <strong>Secure Shell</strong> is a method for secure remote login from one computer to another. </li>
</ul>
<p><strong>X Window System (X11)</strong></p>
<ul>
<li>The X Window System (X11) is an open source, cross platform,  client-server computer software system that provides a <strong>GUI</strong> in a  distributed network environment.</li>
</ul>
<p><strong>Login Hostname</strong></p>
<ul>
<li>login.cluster.tufts.edu = login.pax.tufts.edu</li>
</ul>
<p><strong>Cluster New User Guides</strong></p>
<ul>
<li><a href="https://tufts.box.com/v/HPC-Table-of-Contents">Table of Contents</a></li>
</ul>
</blockquote>
<h2 id="ondemand">OnDemand</h2>
<p>Preferred browser: <strong>Chrome or FireFox</strong></p>
<p>Go to <a href="https://ondemand.pax.tufts.edu"><strong>OnDemand</strong></a>, https://ondemand.pax.tufts.edu</p>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/ondemand_login.png" alt="Core-Node" width=70%></p>
<p>Use your <strong>Tufts UTLN</strong> (Tufts username, lower case!) and <strong>password</strong> to login. </p>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/ondemand_home.png" alt="Core-Node" width=70%></p>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/ondemand_menu.png" alt="Core-Node" width=70%></p>
<p><strong><code>Clusters</code></strong>, you can start a shell access to the HPC cluster. </p>
<p><strong><code>Tufts HPC Shell Access</code></strong> = <code>$ ssh your_utln@login.cluster.tufts.edu</code>= <code>$ ssh your_utln@login.pax.tufts.edu</code></p>
<p>OR</p>
<p>Use the <code>&gt;_Open in Terminal</code> button in <code>Files</code> to open a terminal in whichever directory you navigated to.</p>
<p>If you need X11 access through OnDemand to display any GUI applications, please use our <a href="https://ondemand.pax.tufts.edu">OnDemand</a> <strong><code>Clusters</code></strong> for this option:</p>
<p><strong><code>Tufts HPC FastX11 Shell Access</code></strong> = <code>$ ssh -XYC your_utln@login.cluster.tufts.edu</code> (with X11 for GUI applications)</p>
<p><a href="https://tufts.box.com/s/s1vig4km289dzx8qkq4mbhlp4es0oxu1">FastX Web/Desktop Client Setup Instructions</a></p>
<p>OR </p>
<p>You also have the option to use the <code>Xfce Terminal</code> under new  <a href="https://ondemand.pax.tufts.edu">OnDemand</a> <code>Interactive Apps</code> with limited computing resources.</p>
<h2 id="shell-access">Shell Access</h2>
<blockquote>
<p><strong>Hostname</strong>: <code>login.cluster.tufts.edu</code> or <code>login.pax.tufts.edu</code></p>
</blockquote>
<h3 id="mac-osx-linux">Mac OSX &amp; Linux</h3>
<ul>
<li>
<p><strong>Terminal</strong> </p>
</li>
<li>
<p>Shell environment (default: bash):</p>
<p><code>$ ssh your_utln@login.cluster.tufts.edu</code></p>
<p><code>$ ssh your_utln@login.cluster.tufts.edu</code></p>
<p>With GUI (Graphical User Interface):</p>
<p><code>$ ssh -XC your_utln@login.cluster.tufts.edu</code></p>
<p>or</p>
<p><code>$ ssh -YC your_utln@login.cluster.tufts.edu</code></p>
<p>X Window System need to be locally installed.</p>
<p>Now you are on a <strong>Login Node</strong> of the cluster (login-prod-[01-03]) and in your <strong>Home Directory</strong> (~). </p>
<p><code>$ [your_utln@login-prod-03 ~]</code></p>
</li>
<li>
<p>Setting up <a href="_https://www.tecmint.com/ssh-passwordless-login-using-ssh-keygen-in-5-easy-steps/_">SSH keyless access</a> </p>
<ul>
<li>Be sure your <code>~/.ssh</code> permission is correct! Otherwise, SSH won't work properly.</li>
<li><code>. ssh</code> directory: 700 ( drwx------ )</li>
<li>public key ( <code>. pub</code> file): 644 ( -rw-r--r-- )</li>
<li>private key (<code>id_rsa</code> ): 600 ( -rw------- )</li>
</ul>
</li>
</ul>
<h3 id="windows">Windows</h3>
<ul>
<li><strong><a href="https://learn.microsoft.com/en-us/powershell/scripting/learn/remoting/ssh-remoting-in-powershell-core?view=powershell-7.2">PowerShell</a></strong></li>
<li><strong><a href="https://learn.microsoft.com/en-us/windows/wsl/install">WSL - Windows Subsystem for linux</a></strong></li>
<li><strong><a href="https://www.putty.org/">PuTTY</a></strong>  </li>
<li><strong><a href="https://www.cygwin.com/">Cygwin</a></strong> </li>
</ul>
<p>Need Assistance? Contact us at tts-research@tufts.edu</p>
<h2 id="file-transfers-tofrom-tufts-hpc-cluster">File Transfers To/From Tufts HPC Cluster</h2>
<blockquote>
<p><strong>File Transfer Hostname</strong></p>
<ul>
<li>xfer.cluster.tufts.edu = xfer.pax.tufts.edu</li>
</ul>
<p><strong>File Transfer Protocol</strong></p>
<ul>
<li>SCP</li>
<li>SFTP - Use this for NCBI uploads</li>
<li>rsync over SSH</li>
</ul>
<p><strong>Globus</strong> * - Coming Soon!</p>
<ul>
<li>Tufts HPC Cluster</li>
<li>Tufts Box</li>
<li>Tufts Sharepoint</li>
</ul>
</blockquote>
<h3 id="file-transfer-clients">File Transfer Clients</h3>
<ul>
<li>Windows Only - <strong><a href="https://winscp.net/eng/index.php">WinSCP</a></strong></li>
<li><strong><a href="https://filezilla-project.org/">FileZilla</a></strong>    </li>
<li><strong><a href="https://cyberduck.io/">Cyberduck</a></strong></li>
</ul>
<h3 id="ondemand_1">OnDemand</h3>
<ul>
<li><strong><a href="https://ondemand.pax.tufts.edu">OnDemand</a></strong> (Single file size up to 976MB)</li>
</ul>
<p>Only for transfering files size <strong>less than 976MB per file.</strong></p>
<p>Go to OnDemand:</p>
<p><strong><a href="https://ondemand.pax.tufts.edu/">https://ondemand.pax.tufts.edu/</a></strong> </p>
<p>Under <strong><code>Files</code></strong>, using the <strong><code>Upload</code></strong> or <strong><code>Download</code></strong> buttons to transfer. Make sure you navigate to the destination/source directory on cluster using the <strong><code>Go To</code></strong> button before transfering files.</p>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/ondemand_homedir.png" alt="Core-Node" width=70%></p>
<h3 id="terminal">Terminal</h3>
<ul>
<li><strong>Hostname for file transfer: xfer.cluster.tufts.edu</strong></li>
</ul>
<blockquote>
<p>NOTE:</p>
<ul>
<li><strong>Local_Path</strong> is the path to your files or directory on your local computer</li>
<li><strong>Cluster_Path</strong> is the path to your files or directory on the cluster</li>
<li>Home Directory: <em>/cluster/home/your_utln/your_folder</em></li>
<li>Research Project Storage Space Directory: <em>/cluster/tufts/yourlabname/your_utln/your_folder</em></li>
</ul>
</blockquote>
<p>**<em>Execute from your local machine terminal.</em> **</p>
<p>General Format:</p>
<p><code>$ scp From_Path To_Path</code></p>
<p><code>$ rsync From_Path To_Pat</code></p>
<p><strong><em>NOTE: If you are transfering very large files that could take hours to finish, we would suggest using <code>rsync</code> as it has ability to restart from where it left if interrupted.</em></strong></p>
<p><strong>File</strong> Transfer with <code>scp</code>or <code>rsync</code>:</p>
<ul>
<li>Download from cluster</li>
</ul>
<p><code>$ scp your_utln@xfer.cluster.tufts.edu:Cluster_Path Local_Path</code></p>
<p><code>$ rsync your_utln@xfer.cluster.tufts.edu:Cluster_Path Local_Path</code></p>
<ul>
<li>Upload to cluster</li>
</ul>
<p><code>$ scp Local_Path your_utln@xfer.cluster.tufts.edu:Cluster_Path</code></p>
<p><code>$ rsync Local_Path your_utln@xfer.cluster.tufts.edu:Cluster_Path</code></p>
<p><strong>Directory</strong> Transfer with <code>scp</code> or <code>rsync</code>:</p>
<ul>
<li>Download from cluster</li>
</ul>
<p><code>$ scp -r your_utln@xfer.cluster.tufts.edu:Cluster_Path Local_Path</code></p>
<p><code>$ rsync -azP your_utln@xfer.cluster.tufts.edu:Cluster_Path Local_Path</code></p>
<ul>
<li>Upload to cluster</li>
</ul>
<p><code>$ scp -r Local_Path your_utln@xfer.cluster.tufts.edu:Cluster_Path</code></p>
<p><code>$ rsync -azP Local_Path your_utln@xfer.cluster.tufts.edu:Cluster_Path</code></p>
<h3 id="globus-coming-soon">Globus - Coming Soon!</h3>
<p><a href="https://www.globus.org/">Globus</a> is a research cyberinfrastructure, developed and operated as a not-for-profit service by the University of Chicago.</p>
<p>Local Computer: <a href="https://docs.globus.org/globus-connect-personal/">Globus Connect Personal</a></p>
<p>Collections: </p>
<ul>
<li>
<p>Tufts HPC Cluster Storage (Home and Project)</p>
</li>
<li>
<p>Tufts Box</p>
</li>
<li>
<p>Tufts Sharepoint</p>
</li>
</ul>
<h2 id="software-modules-packages">Software, Modules, Packages</h2>
<h3 id="what-are-modules">What are modules?</h3>
<ul>
<li>A tool that <strong>simplify</strong> shell initialization and lets users easily modify their environment during the session with modulefiles</li>
<li>Each modulefile contains the <strong>information</strong> needed to configure the shell for an application. (PATH, LD_LIBRARY_PATH, CPATH, etc.). Without modules, these environment variables need to be set manually in every new session where the application is needed. </li>
<li>Modules are useful in managing <strong>different versions</strong> of applications. </li>
<li>Modules can also be bundled into metamodules that will load an entire <strong>set of different applications</strong>. </li>
</ul>
<h3 id="whats-new">What's NEW?</h3>
<ul>
<li>Switched from TCL to <a href="https://lmod.readthedocs.io/en/latest/010_user.html">Lmod</a></li>
<li>All previous module commands work as they should + more</li>
<li>Allows module usage tracking</li>
</ul>
<h3 id="how-to-use-modules">How to use modules?</h3>
<blockquote>
<p><strong>Cheat Sheet</strong></p>
<p><code>module av</code> - check available modules on the MODULEPATH</p>
<p><code>module av &lt;software&gt;</code> - check if a specific software is available as a module</p>
<p><code>module spider &lt;keyword&gt;</code> * - lists all possible modules and not just the modules that can be seen in the current MODULEPATH (such as private modules)</p>
<p><code>module --raw show &lt;module_name&gt;</code> * - printing the modulefile</p>
<p><code>module list</code> - check loaded modules</p>
<p><code>module load &lt;software&gt;</code> - load a specific module</p>
<p><code>module unload &lt;software&gt;</code> - unload a specific module</p>
<p><code>module swap &lt;loaded_software&gt; &lt;new_software&gt;</code> - switch a loaded module for a new one</p>
<p><code>module purge</code> - unload all loaded modules</p>
</blockquote>
<p>To <strong>check available modules</strong> installed on the cluster, this may take a few minutes as there are a lot of modules, and be sure to browse the entire list as there are several module file locations:</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ module av
</code></pre></div>
<p>Upon login, environment variable <strong><code>PATH</code></strong> is set for the system to search <strong>executables</strong> along these paths:</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/cluster/home/tutln01/bin:/cluster/home/tutln01/.local/bin
</code></pre></div>
<p>For example, I would like to use <code>gcc</code> compiler, to check what versions of gcc compiler is available, load the version I would like to use, and use it:</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ module av gcc

----------------------------------------------------------- /opt/shared/Modules/modulefiles-rhel6 ------------------------------------------------------------
gcc/4.7.0 gcc/4.9.2 gcc/5.3.0 gcc/7.3.0

-------------------------------------------------------------- /cluster/tufts/hpc/tools/module ---------------------------------------------------------------
gcc/8.4.0 gcc/9.3.0 gcc/11.2.0
</code></pre></div>
<p>Use <code>module list</code> to <strong>check loaded modules</strong> in current environment:</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ module load gcc/7.3.0
[tutln01@login-prod-01 ~]$ module list
Currently Loaded Modulefiles:
  1) use.own     2) gcc/7.3.0
</code></pre></div>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ which gcc
/opt/shared/gcc/7.3.0/bin/gcc
[tutln01@login-prod-01 ~]$ echo $PATH
/opt/shared/gcc/7.3.0/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/cluster/home/tutln01/bin:/cluster/home/tutln01/.local/bin
[tutln01@login-prod-01 ~]$ gcc --version
gcc (GCC) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
</code></pre></div>
<p><strong>swap a module for another</strong> (doesn't have to be the same software):</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ module swap gcc/7.3.0 gcc/9.3.0 
[tutln01@login-prod-01 ~]$ module list
Currently Loaded Modulefiles:
  1) use.own     2) gcc/9.3.0
</code></pre></div>
<p><strong>unload loaded modules</strong>:</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ module unload gcc
[tutln01@login-prod-01 ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/cluster/home/tutln01/bin:/cluster/home/tutln01/.local/bin
</code></pre></div>
<p><strong>unload ALL</strong> of the loaded modules in the current environment:</p>
<div class="highlight"><pre><span></span><code>[tutln01@login-prod-01 ~]$ module purge
</code></pre></div>
<h3 id="install-softwarepackages">Install Software/Packages</h3>
<ul>
<li><a href="https://tufts.box.com/s/qximkv5ke2y4k0vbg6m04m6fc6exh88h">R</a> (R command line recommanded)</li>
<li>Packages need to be <u>reinstalled</u> for each version of R</li>
<li>OnDemand <ul>
<li>Interactive Apps - RStudio Pax</li>
<li>Bioinformatics Apps - RStudio for bioinformatics, RStudio for scRNA-Seq</li>
</ul>
</li>
<li><a href="https://tuftsdatalab.github.io/tuftsWorkshops/2024_workshops/2024_bioinformatics301/03_conda/">Python</a> (Conda env recommanded)</li>
<li>Not recommended: anaconda/3 or anaconda/2 (older versions), only use this version when it's absolutely necessary</li>
<li>Other software compiled from <strong>source</strong></li>
<li>gcc</li>
<li>cmake</li>
<li>Autotools - automake, autoconf, autogen, .etc</li>
<li>... any <strong>dependencies</strong>, load if available, install if not. Some environment variables may need to be set manually.</li>
<li><strong>Follow instructions</strong> (read it through first)</li>
<li>Use <strong>"--prefix="</strong> to install in non-standard locations</li>
<li>Modify the environment variables !!! (such as PATH, LD_LIBRARY_PATH, CPATH, .etc)</li>
<li>You can make a private module for your locally installed software. Here is <a href="https://tufts.box.com/s/fewt978g2hmmmhskm1n5dg4bj6xt0cdd">HOW</a></li>
<li>OR you can submit a request to tts-research@tufts.edu for the software to be installed globally on HPC cluster to share with the community. </li>
</ul>
<hr />
<h1 id="getting-work-done-on-tufts-hpc-cluster-with-slurm">Getting Work Done on Tufts HPC Cluster with SLURM</h1>
<p><img src="https://ormstonhouse.com/wp-content/uploads/2019/09/tetris.jpg" alt="teris" width=50%></p>
<p><strong>ALL</strong> work <strong>MUST</strong> to be performed on <strong>compute nodes</strong>!</p>
<p>If you see prompt like this <code>[your_utln@login-prod-01]</code>, <code>[your_utln@login-prod-02]</code>, <code>[your_utln@login-prod-03]</code>, <strong>DON'T</strong> run any programs! <strong>Get resource allocation first</strong>!</p>
<p><img src="https://raw.githubusercontent.com/DelilahYM/ImageHost/master/Cluster_20230516.png" alt="Memory-Storage" width=70%></p>
<blockquote>
<p><strong>Things to think about before requesting resources:</strong></p>
<ul>
<li>What program?</li>
<li>What kind of resources? CPUs only or with GPUs?</li>
<li>How much memory?</li>
<li>How many CPU cores?</li>
<li>How long does the job need to run?</li>
<li>Command line application? GUI?</li>
<li>Does the program need interactions?</li>
<li>Prototyping, debugging, or production?</li>
</ul>
</blockquote>
<h2 id="ondemand_2">OnDemand</h2>
<ul>
<li>
<p><strong><a href="https://ondemand.pax.tufts.edu/">OnDemand (https://ondemand.pax.tufts.edu)</a></strong></p>
</li>
<li>
<p><strong><code>Interactive Apps</code></strong> --&gt; RStudio, Matlab, JupyterLab, Jupyter Notebook, .etc</p>
</li>
<li>
<p><strong><code>Clusters</code></strong> --&gt; Tufts HPC Cluster  Shell Access </p>
</li>
<li>
<p><strong><code>Files</code></strong> </p>
</li>
<li>
<p><strong><code>Jobs</code></strong></p>
</li>
<li>
<p><a href="https://ondemand.pax.tufts.edu/">OnDemand</a>  <strong><code>Clusters</code></strong> --&gt; Tufts HPC Cluster FastX11 Shell Access</p>
</li>
</ul>
<h2 id="slurm-information">Slurm Information</h2>
<ul>
<li>View information about Slurm nodes and partitions.</li>
</ul>
<p><code>$ sinfo</code></p>
<p>With more specifc information and formated output:</p>
<p><code>$ sinfo -o "%20N %10P %10c %10m %85f %10G "</code> - NODELIST, PARTITION, CPUS, MEMORY,AVAIL_FEATURES, GRES  </p>
<p>More  <a href="https://slurm.schedmd.com/sinfo.html">sinfo</a> options</p>
<p><strong>You can only see the partitions you have access to.</strong></p>
<p><strong>For most users</strong>, you will see <code>batch</code>,<code>mpi</code>,<code>gpu</code>,<code>largemem</code>, and <code>preempt</code> partitions.          </p>
<p>How to check <strong>GPU, Memory, CPU</strong> availability on the cluster?</p>
<h2 id="utilize-new-hpctools-module">Utilize New <code>hpctools</code> Module !!!</h2>
<p>Users can use <code>hpctools</code> module to check: **Free CPU resources, Free GPU resources, User Past and Active jobs, and Project space quota and usage. **</p>
<p>```[tutln01@login-prod-01 ~]$ module load hpctools
[tutln01@login-prod-01 ~]$ module load hpctools
     command: hpctools
[tutln01@login-prod-01 ~]$ hpctools
 Please select from the following options:</p>
<ol>
<li>
<p>Checking Free Resources On Each Node in Given Partition(s)</p>
</li>
<li>
<p>Checking Free GPU Resources On Each Node in Given Partition(s)</p>
</li>
<li>
<p>Checking tutln01 Past Completed Jobs in Given Time Period</p>
</li>
<li>
<p>Checking tutln01 Active Job informantion</p>
</li>
<li>
<p>Checking Project Space Storage Quota Informantion</p>
</li>
<li>
<p>Checking Any Directory Storage Usage Informantion</p>
</li>
</ol>
<p>Press q to quit</p>
<p>Your Selection:</p>
<div class="highlight"><pre><span></span><code>Then follow the onscreen instructions to get desired information.



## Interactive Session

- Particularly good for debugging and working with software GUI. 

  `$ srun [options] --pty [command]`

- Command 

  - command to run an application, given the module is already loaded.
  - `bash` for a bash shell

- Options

  - Pseudo terminal `--pty`
  - Partition `-p` or `--partition=`
    - Default batch if not specified
  - Time `-t` or `--time=`
    - Default 15 minutes if not specified on non-interactive partition
    - Format: DD-HH:MM:SS
  - Number of CPU cores `-n` 
    - Default 1 if not specified
  - Memory `--mem=`
    - Default 2GB if not specified
  - GPU `--gres=`
    - Default none
  - Features `--constraint=`
    - GPU types
    - OS version
    - CPU architecture
    - Instruction Set
    - Default none
  - X Window `--x11=first`
    - Default none  

**Starting an interactive session of bash shell on preempt partition with 2 CPU cores and 2GB of RAM, with X11 forwarding for 1 day, 2 hours, and 30 minutes (use `exit` to end session and release resources).**

```bash
[tutln01@login-prod-01 ~]$ srun -p preempt -t 1-2:30:00 -n 2 --mem=2g --x11=first --pty bash
srun: job 296794 queued and waiting for resources
srun: job 296794 has been allocated resources
[tutln01@cc1gpu001 ~]$ 
</code></pre></div>
<p>Note: If you are requesting X11 forwarding with <code>srun</code>, <code>-XC</code> or<code>-YC</code> or <code>-XYC</code> must be used upon login with <code>ssh</code>.</p>
<p><strong>Starting an interactive session of bash shell on preempt partition with 2 CPU cores and 4GB of RAM, with 1 A100 GPU for 1 day, 2 hours, and 30 minutes (use <code>exit</code> to end session and release resources).</strong></p>
<div class="highlight"><pre><span></span><code><span class="o">[</span>tutln01@login-prod-01<span class="w"> </span>~<span class="o">]</span>$<span class="w"> </span>srun<span class="w"> </span>-p<span class="w"> </span>preempt<span class="w"> </span>-t<span class="w"> </span><span class="m">1</span>-2:30:00<span class="w"> </span>-n<span class="w"> </span><span class="m">2</span><span class="w"> </span>--mem<span class="o">=</span>4g<span class="w"> </span>--gres<span class="o">=</span>gpu:a100:1<span class="w"> </span>--pty<span class="w"> </span>bash
</code></pre></div>
<p>Once your resource is allocated on a compute node, use <code>nvidia-smi</code> to check GPU info.</p>
<h2 id="batch-job">Batch Job</h2>
<p>Write a batch submission script e.g. <strong>myjob.sh</strong></p>
<div class="highlight"><pre><span></span><code><span class="ch">#!/bin/bash -l</span>
<span class="c1">#SBATCH -J My_Job_Name   #job name</span>
<span class="c1">#SBATCH --time=00-00:20:00  #requested time (DD-HH:MM:SS)</span>
<span class="c1">#SBATCH -p batch,preempt    #running on &quot;batch&quot; or &quot;preempt&quot; partition, wherever resource is available first</span>
<span class="c1">#SBATCH -N 1    #1 nodes #for many shared-memory programs,please leave -N as 1.</span>
<span class="c1">#SBATCH -n 2   #2 tasks total and 1 cpu per task, that gives you 2 cpu cores for this job</span>
<span class="c1">#SBATCH --mem=2g  #requesting 2GB of RAM total for the number of cpus you requested</span>
<span class="c1">##SBATCH --gres=gpu:a100:1  #requesting 1 A100 GPU, in this case, the &quot;-p&quot; needs to be switched to a partition has the requested GPU resources</span>
<span class="c1">#SBATCH --output=MyJob.%j.%N.out  #saving standard output to file, %j=JOBID, %N=NodeName</span>
<span class="c1">#SBATCH --error=MyJob.%j.%N.err   #saving standard error to file, %j=JOBID, %N=NodeName</span>
<span class="c1">#SBATCH --mail-type=ALL    #email optitions</span>
<span class="c1">#SBATCH --mail-user=Your_Tufts_Email@tufts.edu</span>

<span class="c1">#[commands_you_would_like_to_exe_on_the_compute_nodes]</span>
<span class="c1"># unload all modules to ensure a clean start.</span>
module<span class="w"> </span>purge
<span class="c1"># for example, running a python script </span>
<span class="c1"># load the module so the correct version python is available to you</span>
module<span class="w"> </span>load<span class="w"> </span>anaconda/2021.05
<span class="c1"># If you have a conda env that you would like to use, activate it here using &quot;source activate xxx&quot;. DO NOT USE &quot;conda activate&quot;</span>
<span class="nb">source</span><span class="w"> </span>activate<span class="w"> </span><span class="o">[</span>target_env<span class="o">]</span>
<span class="c1"># run python script</span>
python<span class="w"> </span>myscript.py<span class="w"> </span><span class="c1">#make sure myscript.py exists in the current directory</span>
<span class="c1"># make sure you save all plots, data, outputs generated to files in your script</span>
<span class="c1"># Don&#39;t forget to deactivate your conda env if you are using one</span>
conda<span class="w"> </span>deactivate
</code></pre></div>
<p><strong>Submit</strong> the job using the following command from command line interface:</p>
<p><code>$ sbatch myjob.sh</code></p>
<p><strong>Sample Scripts</strong> including R, conda, matlab, gaussian</p>
<p><code>/cluster/tufts/hpc/tools/slurm_scripts</code></p>
<h2 id="job-status">Job Status</h2>
<ul>
<li>Checking your <strong>active</strong> jobs</li>
</ul>
<div class="highlight"><pre><span></span><code><span class="o">[</span>tutln01@cc1gpu001<span class="w"> </span>~<span class="o">]</span>$<span class="w"> </span>squeue<span class="w"> </span>-u<span class="w"> </span><span class="nv">$USER</span>
<span class="w">             </span>JOBID<span class="w"> </span>PARTITION<span class="w">     </span>NAME<span class="w">     </span>USER<span class="w"> </span>ST<span class="w">       </span>TIME<span class="w">  </span>NODES<span class="w"> </span>NODELIST<span class="o">(</span>REASON<span class="o">)</span><span class="w"> </span>
<span class="w">            </span><span class="m">296794</span><span class="w">   </span>preempt<span class="w">     </span>bash<span class="w"> </span>tutln01<span class="w">  </span>R<span class="w">       </span><span class="m">5</span>:12<span class="w">      </span><span class="m">1</span><span class="w"> </span>cc1gpu001<span class="w"> </span>
<span class="o">[</span>tutln01@cc1gpu001<span class="w"> </span>~<span class="o">]</span>$<span class="w"> </span>squeue<span class="w"> </span>-u<span class="w"> </span>tutln01
<span class="w">             </span>JOBID<span class="w"> </span>PARTITION<span class="w">     </span>NAME<span class="w">     </span>USER<span class="w"> </span>ST<span class="w">       </span>TIME<span class="w">  </span>NODES<span class="w"> </span>NODELIST<span class="o">(</span>REASON<span class="o">)</span><span class="w"> </span>
<span class="w">            </span><span class="m">296794</span><span class="w">   </span>preempt<span class="w">     </span>bash<span class="w"> </span>tutln01<span class="w">  </span>R<span class="w">       </span><span class="m">5</span>:21<span class="w">      </span><span class="m">1</span><span class="w"> </span>cc1gpu001<span class="w"> </span>
</code></pre></div>
<p>To check your active jobs in the queue:</p>
<p><code>$ squeue -u $USER</code> or <code>$ squeue -u your_utln</code></p>
<p>To cancel a specific job:</p>
<p><code>$ scancel JOBID</code></p>
<p>To cancel all of your jobs:</p>
<p><code>$ scancel -u $USER</code> or <code>$ scancel -u your_utln</code></p>
<p>To check details of your <strong>active jobs</strong> (running or pending):</p>
<p><code>$ scontrol show jobid -dd JOBID</code></p>
<div class="highlight"><pre><span></span><code><span class="o">[</span>tutln01@cc1gpu001<span class="w"> </span>~<span class="o">]</span>$<span class="w"> </span>scontrol<span class="w"> </span>show<span class="w"> </span>jobid<span class="w"> </span>-dd<span class="w"> </span><span class="m">296794</span>
<span class="nv">JobId</span><span class="o">=</span><span class="m">296794</span><span class="w"> </span><span class="nv">JobName</span><span class="o">=</span>bash
<span class="w">   </span><span class="nv">UserId</span><span class="o">=</span>tutln01<span class="o">(</span><span class="m">31003</span><span class="o">)</span><span class="w"> </span><span class="nv">GroupId</span><span class="o">=</span>tutln01<span class="o">(</span><span class="m">5343</span><span class="o">)</span><span class="w"> </span><span class="nv">MCS_label</span><span class="o">=</span>N/A
<span class="w">   </span><span class="nv">Priority</span><span class="o">=</span><span class="m">10833</span><span class="w"> </span><span class="nv">Nice</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">Account</span><span class="o">=(</span>null<span class="o">)</span><span class="w"> </span><span class="nv">QOS</span><span class="o">=</span>normal
<span class="w">   </span><span class="nv">JobState</span><span class="o">=</span>RUNNING<span class="w"> </span><span class="nv">Reason</span><span class="o">=</span>None<span class="w"> </span><span class="nv">Dependency</span><span class="o">=(</span>null<span class="o">)</span>
<span class="w">   </span><span class="nv">Requeue</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">Restarts</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">BatchFlag</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">Reboot</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">ExitCode</span><span class="o">=</span><span class="m">0</span>:0
<span class="w">   </span><span class="nv">DerivedExitCode</span><span class="o">=</span><span class="m">0</span>:0
<span class="w">   </span><span class="nv">RunTime</span><span class="o">=</span><span class="m">00</span>:10:33<span class="w"> </span><span class="nv">TimeLimit</span><span class="o">=</span><span class="m">1</span>-02:30:00<span class="w"> </span><span class="nv">TimeMin</span><span class="o">=</span>N/A
<span class="w">   </span><span class="nv">SubmitTime</span><span class="o">=</span><span class="m">2021</span>-03-22T22:18:50<span class="w"> </span><span class="nv">EligibleTime</span><span class="o">=</span><span class="m">2021</span>-03-22T22:18:50
<span class="w">   </span><span class="nv">AccrueTime</span><span class="o">=</span><span class="m">2021</span>-03-22T22:18:50
<span class="w">   </span><span class="nv">StartTime</span><span class="o">=</span><span class="m">2021</span>-03-22T22:18:55<span class="w"> </span><span class="nv">EndTime</span><span class="o">=</span><span class="m">2021</span>-03-24T00:48:55<span class="w"> </span><span class="nv">Deadline</span><span class="o">=</span>N/A
<span class="w">   </span><span class="nv">PreemptEligibleTime</span><span class="o">=</span><span class="m">2021</span>-03-22T22:18:55<span class="w"> </span><span class="nv">PreemptTime</span><span class="o">=</span>None
<span class="w">   </span><span class="nv">SuspendTime</span><span class="o">=</span>None<span class="w"> </span><span class="nv">SecsPreSuspend</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">LastSchedEval</span><span class="o">=</span><span class="m">2021</span>-03-22T22:18:55
<span class="w">   </span><span class="nv">Partition</span><span class="o">=</span>preempt<span class="w"> </span>AllocNode:Sid<span class="o">=</span>login-prod-01:34458
<span class="w">   </span><span class="nv">ReqNodeList</span><span class="o">=(</span>null<span class="o">)</span><span class="w"> </span><span class="nv">ExcNodeList</span><span class="o">=(</span>null<span class="o">)</span>
<span class="w">   </span><span class="nv">NodeList</span><span class="o">=</span>cc1gpu001
<span class="w">   </span><span class="nv">BatchHost</span><span class="o">=</span>cc1gpu001
<span class="w">   </span><span class="nv">NumNodes</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="nv">NumCPUs</span><span class="o">=</span><span class="m">2</span><span class="w"> </span><span class="nv">NumTasks</span><span class="o">=</span><span class="m">2</span><span class="w"> </span>CPUs/Task<span class="o">=</span><span class="m">1</span><span class="w"> </span>ReqB:S:C:T<span class="o">=</span><span class="m">0</span>:0:*:*
<span class="w">   </span><span class="nv">TRES</span><span class="o">=</span><span class="nv">cpu</span><span class="o">=</span><span class="m">2</span>,mem<span class="o">=</span>2G,node<span class="o">=</span><span class="m">1</span>,billing<span class="o">=</span><span class="m">2</span>
<span class="w">   </span>Socks/Node<span class="o">=</span>*<span class="w"> </span>NtasksPerN:B:S:C<span class="o">=</span><span class="m">0</span>:0:*:*<span class="w"> </span><span class="nv">CoreSpec</span><span class="o">=</span>*
<span class="w">   </span><span class="nv">JOB_GRES</span><span class="o">=(</span>null<span class="o">)</span>
<span class="w">     </span><span class="nv">Nodes</span><span class="o">=</span>cc1gpu001<span class="w"> </span><span class="nv">CPU_IDs</span><span class="o">=</span><span class="m">30</span>-31<span class="w"> </span><span class="nv">Mem</span><span class="o">=</span><span class="m">2048</span><span class="w"> </span><span class="nv">GRES</span><span class="o">=</span>
<span class="w">   </span><span class="nv">MinCPUsNode</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="nv">MinMemoryNode</span><span class="o">=</span>2G<span class="w"> </span><span class="nv">MinTmpDiskNode</span><span class="o">=</span><span class="m">0</span>
<span class="w">   </span><span class="nv">Features</span><span class="o">=(</span>null<span class="o">)</span><span class="w"> </span><span class="nv">DelayBoot</span><span class="o">=</span><span class="m">00</span>:00:00
<span class="w">   </span><span class="nv">OverSubscribe</span><span class="o">=</span>OK<span class="w"> </span><span class="nv">Contiguous</span><span class="o">=</span><span class="m">0</span><span class="w"> </span><span class="nv">Licenses</span><span class="o">=(</span>null<span class="o">)</span><span class="w"> </span><span class="nv">Network</span><span class="o">=(</span>null<span class="o">)</span>
<span class="w">   </span><span class="nv">Command</span><span class="o">=</span>bash
<span class="w">   </span><span class="nv">WorkDir</span><span class="o">=</span>/cluster/home/tutln01
<span class="w">   </span><span class="nv">Power</span><span class="o">=</span>
<span class="w">   </span><span class="nv">MailUser</span><span class="o">=</span>tutln01<span class="w"> </span><span class="nv">MailType</span><span class="o">=</span>NONE
</code></pre></div>
<ul>
<li>Checking your <strong>finished</strong> jobs</li>
</ul>
<p><em>You can no longer see these jobs in <code>squeue</code> command output.</em></p>
<p>Querying finished jobs helps users make better decisions on requesting resources for future jobs. </p>
<p><strong>Check job CPU, memory usage, and efficiency:</strong></p>
<p><code>$ seff JOBID</code></p>
<div class="highlight"><pre><span></span><code><span class="o">[</span>tutln01@login-prod-01<span class="w"> </span>~<span class="o">]</span>$<span class="w"> </span>seff<span class="w"> </span><span class="m">296794</span>
Job<span class="w"> </span>ID:<span class="w"> </span><span class="m">296794</span>
Cluster:<span class="w"> </span>pax
Use<span class="w"> </span>of<span class="w"> </span>uninitialized<span class="w"> </span>value<span class="w"> </span><span class="nv">$user</span><span class="w"> </span><span class="k">in</span><span class="w"> </span>concatenation<span class="w"> </span><span class="o">(</span>.<span class="o">)</span><span class="w"> </span>or<span class="w"> </span>string<span class="w"> </span>at<span class="w"> </span>/usr/bin/seff<span class="w"> </span>line<span class="w"> </span><span class="m">154</span>,<span class="w"> </span>&lt;DATA&gt;<span class="w"> </span>line<span class="w"> </span><span class="m">602</span>.
User/Group:<span class="w"> </span>/tutln01
State:<span class="w"> </span>COMPLETED<span class="w"> </span><span class="o">(</span><span class="nb">exit</span><span class="w"> </span>code<span class="w"> </span><span class="m">0</span><span class="o">)</span>
Nodes:<span class="w"> </span><span class="m">1</span>
Cores<span class="w"> </span>per<span class="w"> </span>node:<span class="w"> </span><span class="m">2</span>
CPU<span class="w"> </span>Utilized:<span class="w"> </span><span class="m">00</span>:00:00
CPU<span class="w"> </span>Efficiency:<span class="w"> </span><span class="m">0</span>.00%<span class="w"> </span>of<span class="w"> </span><span class="m">00</span>:22:12<span class="w"> </span>core-walltime
Job<span class="w"> </span>Wall-clock<span class="w"> </span>time:<span class="w"> </span><span class="m">00</span>:11:06
Memory<span class="w"> </span>Utilized:<span class="w"> </span><span class="m">1</span>.16<span class="w"> </span>MB<span class="w"> </span><span class="o">(</span>estimated<span class="w"> </span>maximum<span class="o">)</span>
Memory<span class="w"> </span>Efficiency:<span class="w"> </span><span class="m">0</span>.06%<span class="w"> </span>of<span class="w"> </span><span class="m">2</span>.00<span class="w"> </span>GB<span class="w"> </span><span class="o">(</span><span class="m">2</span>.00<span class="w"> </span>GB/node<span class="o">)</span>
</code></pre></div>
<p><strong>Check job detailed accounting data:</strong></p>
<p><code>$ sacct --format=partition,state,time,start,end,elapsed,MaxRss,ReqMem,MaxVMSize,nnodes,ncpus,nodelist -j JOBID</code></p>
<div class="highlight"><pre><span></span><code><span class="o">[</span>tutln01@login-prod-01<span class="w"> </span>~<span class="o">]</span>$<span class="w"> </span>sacct<span class="w"> </span>--format<span class="o">=</span>partition,state,time,start,end,elapsed,MaxRss,ReqMem,MaxVMSize,nnodes,ncpus,nodelist<span class="w"> </span>-j<span class="w">  </span><span class="m">296794</span>
<span class="w"> </span>Partition<span class="w">      </span>State<span class="w">  </span>Timelimit<span class="w">               </span>Start<span class="w">                 </span>End<span class="w">    </span>Elapsed<span class="w">     </span>MaxRSS<span class="w">     </span>ReqMem<span class="w">  </span>MaxVMSize<span class="w">   </span>NNodes<span class="w">      </span>NCPUS<span class="w">        </span>NodeList<span class="w"> </span>
----------<span class="w"> </span>----------<span class="w"> </span>----------<span class="w"> </span>-------------------<span class="w"> </span>-------------------<span class="w"> </span>----------<span class="w"> </span>----------<span class="w"> </span>----------<span class="w"> </span>----------<span class="w"> </span>--------<span class="w"> </span>----------<span class="w"> </span>---------------<span class="w"> </span>
<span class="w">   </span>preempt<span class="w">  </span>COMPLETED<span class="w"> </span><span class="m">1</span>-02:30:00<span class="w"> </span><span class="m">2021</span>-03-22T22:18:55<span class="w"> </span><span class="m">2021</span>-03-22T22:30:01<span class="w">   </span><span class="m">00</span>:11:06<span class="w">                   </span>2Gn<span class="w">                   </span><span class="m">1</span><span class="w">          </span><span class="m">2</span><span class="w">       </span>cc1gpu001<span class="w"> </span>
<span class="w">           </span>OUT_OF_ME+<span class="w">            </span><span class="m">2021</span>-03-22T22:18:55<span class="w"> </span><span class="m">2021</span>-03-22T22:30:01<span class="w">   </span><span class="m">00</span>:11:06<span class="w">         </span>8K<span class="w">        </span>2Gn<span class="w">    </span>135100K<span class="w">        </span><span class="m">1</span><span class="w">          </span><span class="m">2</span><span class="w">       </span>cc1gpu001<span class="w"> </span>
<span class="w">            </span>COMPLETED<span class="w">            </span><span class="m">2021</span>-03-22T22:18:56<span class="w"> </span><span class="m">2021</span>-03-22T22:30:01<span class="w">   </span><span class="m">00</span>:11:05<span class="w">       </span>592K<span class="w">        </span>2Gn<span class="w">    </span>351672K<span class="w">        </span><span class="m">1</span><span class="w">          </span><span class="m">2</span><span class="w">       </span>cc1gpu001<span class="w"> </span>
</code></pre></div>
<p>NOTE: there are more format options, see <a href="https://slurm.schedmd.com/sacct.html">sacct</a></p>
<p><strong>OR</strong> utilize <code>hpctools</code> module on the cluster to make things a little easier</p>
<p>```[tutln01@login-prod-01 ~]$ module load hpctools
[tutln01@login-prod-01 ~]$ module load hpctools
     command: hpctools
[tutln01@login-prod-01 ~]$ hpctools
 Please select from the following options:</p>
<ol>
<li>
<p>Checking Free Resources On Each Node in Given Partition(s)</p>
</li>
<li>
<p>Checking Free GPU Resources On Each Node in Given Partition(s)</p>
</li>
<li>
<p>Checking tutln01 Past Completed Jobs in Given Time Period</p>
</li>
<li>
<p>Checking tutln01 Active Job informantion</p>
</li>
<li>
<p>Checking Project Space Storage Quota Informantion</p>
</li>
<li>
<p>Checking Any Directory Storage Usage Informantion</p>
</li>
</ol>
<p>Press q to quit</p>
<p>Your Selection:</p>
<p>```</p>
<p>Then follow the onscreen instructions to get desired information.</p>
<p><strong>Useful Link</strong></p>
<p><a href="https://tufts.box.com/v/HPC-Table-of-Contents">https://tufts.box.com/v/HPC-Table-of-Contents</a></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "content.action.edit", "navigation.footer", "content.code.copy"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.83f73b43.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>